presto-orc/src/main/java/com/facebook/presto/orc/writer/SegmentedSliceBlockBuilder.java

```diff
@@ -0,0 +1,415 @@
+ * they are segmented.
+ * <p>
+ * How is it implemented ?
+ * The Strings from 0 to SEGMENT_SIZE-1 are stored in the first segment.
+ * The string from SEGMENT_SIZE to 2 * SEGMENT_SIZE -1 goes to the second.
+ * Each segment has Slice(data is concatenated and stored in one slice)
+ * and offsets to capture the start offset and length. New slices are appended
+ * to the open segment. Once the segment is full the segment is
+ * finalized and appended to the closed segments. A new open segment is
+ * created for further appends.
+ */
+public class SegmentedSliceBlockBuilder
+        extends AbstractVariableWidthBlock
+        implements BlockBuilder
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(SegmentedSliceBlockBuilder.class).instanceSize();
+
+    private final DynamicSliceOutput openSliceOutput;
+
+    private int openSegmentIndex;
+    private int openSegmentOffset;
+    private int[][] offsets;
+    private Slice[] closedSlices;
+    private long closedSlicesRetainedSize;
+    private long closedSlicesSizeInBytes;
+
+    public SegmentedSliceBlockBuilder(int expectedEntries, int expectedBytes)
+    {
+        int initialSize = Math.max(Segments.INITIAL_SEGMENTS, Segments.segment(expectedEntries) + 1);
+        offsets = new int[initialSize][];
+        closedSlices = new Slice[initialSize];
+        offsets[0] = new int[Segments.SEGMENT_SIZE + 1];
+        openSliceOutput = new DynamicSliceOutput(expectedBytes);
+    }
+
+    public void reset()
+    {
+        openSliceOutput.reset();
+
+        Arrays.fill(closedSlices, null);
+        closedSlicesRetainedSize = 0;
+        closedSlicesSizeInBytes = 0;
+
+        // Fill the first offset array with 0, and free up the rest of the offsets array.
+        Arrays.fill(offsets[0], 0);
+        Arrays.fill(offsets, 1, offsets.length, null);
+        openSegmentIndex = 0;
+        openSegmentOffset = 0;
+    }
+
+    @Override
+    public int getPositionOffset(int position)
+    {
+        return getOffset(position);
+    }
+
+    @Override
+    public int getSliceLength(int position)
+    {
+        int offset = Segments.offset(position);
+        int segment = Segments.segment(position);
+        return offsets[segment][offset + 1] - offsets[segment][offset];
+    }
+
+    private Slice getSegmentRawSlice(int segment)
+    {
+        return segment == openSegmentIndex ? openSliceOutput.getUnderlyingSlice() : closedSlices[segment];
+    }
+
+    @Override
+    public Slice getRawSlice(int position)
+    {
+        return getSegmentRawSlice(Segments.segment(position));
+    }
+
+    @Override
+    public int getPositionCount()
+    {
+        return Segments.getPositions(openSegmentIndex, openSegmentOffset);
+    }
+
+    @Override
+    public long getSizeInBytes()
+    {
+        long offsetsSizeInBytes = Integer.BYTES * (long) getPositionCount();
+        return openSliceOutput.size() + offsetsSizeInBytes + closedSlicesSizeInBytes;
```

Comment:
You need to add the INSTANCE_SIZE too.

Reply:
This will be a deviation from the other getSizeInBytes for Blocks and BlockBuilders. getSizeInBytes, return the size of the data it is holding. getRetainedSize in bytes,  return the size of the memory consumed by this object so need to include the retained SIZE.